packages/editor/src/components/post-actions/set-as-homepage.js

```diff
@@ -0,0 +1,245 @@
+							// translators: %s: title of the page to be set as the homepage.
+							__(
+								'"%s" is a draft and will be published automatically if set as the homepage. Publish page and set as the site homepage?'
+							),
+							pageTitle
+						) }
+					</Text>
+				</>
+			);
+		}
+
+		const modalTranslatedString =
+			// translators: %1$s: title of page to be set as the home page. %2$s: title of the current home page.
+			__(
+				'Set "%1$s" as the site homepage? This will replace the current homepage: "%2$s"'
+			);
+
+		return (
+			<Text>
+				{ sprintf(
+					modalTranslatedString,
+					pageTitle,
+					currentHomePageTitle
+				) }{ ' ' }
+			</Text>
+		);
+	};
+
+	return (
+		<form onSubmit={ onSetPageAsHomepage }>
+			<VStack spacing="5">
+				{ renderModalBody() }
+				<HStack justify="right">
+					<Button
+						__next40pxDefaultSize
+						variant="tertiary"
+						onClick={ () => {
+							closeModal?.();
+						} }
+					>
+						{ __( 'Cancel' ) }
+					</Button>
+					<Button
+						__next40pxDefaultSize
+						variant="primary"
+						type="submit"
+					>
+						{
+							// translators: Button to confirm setting the specified page as the homepage.
+							isPageDraft
+								? __( 'Publish and set homepage' )
+								: __( 'Set homepage' )
+						}
+					</Button>
+				</HStack>
+			</VStack>
+		</form>
+	);
+};
+
+export const useSetAsHomepageAction = () => {
+	const {
+		canManageOptions,
+		hasFrontPageTemplate,
+		pageOnFront,
+		pageForPosts,
+	} = useSelect( ( select ) => {
+		const { getEntityRecord, getEntityRecords } = select( coreStore );
+		const siteSettings = getEntityRecord( 'root', 'site', undefined );
+		const templates = getEntityRecords( 'postType', 'wp_template', {
+			per_page: -1,
+		} );
+
+		return {
+			pageOnFront: siteSettings?.page_on_front,
+			pageForPosts: siteSettings?.page_for_posts,
+			canManageOptions: select( coreStore ).canUser( 'update', {
+				kind: 'root',
+				name: 'site',
+			} ),
+			hasFrontPageTemplate: !! templates?.find(
+				( template ) =>
+					'slug' in template && template.slug === 'front-page'
+			),
+		};
+	} );
+
+	return useMemo(
+		() => ( {
+			id: 'set-as-homepage',
+			label: __( 'Set as homepage' ),
+			isEligible( post ) {
+				if ( post.status === 'trash' ) {
+					return false;
+				}
+
+				if ( post.type !== 'page' ) {
+					return false;
+				}
+
+				// Don't show the action if the page is already set as the homepage.
+				if ( pageOnFront === post.id ) {
+					return false;
+				}
+
+				// Don't show the action if the page is already set as the page for posts.
+				if ( pageForPosts === post.id ) {
+					return false;
+				}
+
+				// Don't show the action if the user can't manage site options.
+				if ( ! canManageOptions ) {
+					return false;
+				}
+
+				// A front-page template overrides homepage settings,
+				// so don't show the setAsHomepage action if it's present.
+				if ( hasFrontPageTemplate ) {
+					return false;
+				}
```

Comment:
These two checks don't use the item. Instead of running them for every item, we can check these when registering the action.

Reply:
Ah I meant to come back to this, thank you. I wasn't sure how to handle this initially but I think I've managed to add a conditional statement here, where the action is registered: https://github.com/WordPress/gutenberg/pull/65426/commits/4cd6988031abb6806396b493eac13c6a62d987d6.